llvm/test/CodeGen/AArch64/aarch64-dup-extract-scalable.ll

```diff
@@ -56,19 +151,133 @@ define <vscale x 8 x half> @dup_extract_f16(<vscale x 8 x half> %data) {
+  %.splat = shufflevector <vscale x 8 x half> %.splatinsert, <vscale x 8 x half> poison, <vscale x 8 x i32> zeroinitializer
+  ret <vscale x 8 x half> %.splat
+}
+
+define <vscale x 8 x half> @dup_extract_nxv8f16_nxv2f16(<vscale x 2 x half> %data) {
+; CHECK-LABEL: dup_extract_nxv8f16_nxv2f16:
+; CHECK:       // %bb.0:
+; CHECK-NEXT:    mov z0.d, z0.d[1]
+; CHECK-NEXT:    ret
+  %1 = extractelement <vscale x 2 x half> %data, i16 1
+  %.splatinsert = insertelement <vscale x 8 x half> poison, half %1, i32 0
+  %.splat = shufflevector <vscale x 8 x half> %.splatinsert, <vscale x 8 x half> poison, <vscale x 8 x i32> zeroinitializer
+  ret <vscale x 8 x half> %.splat
+}
+
+define <vscale x 8 x half> @dup_extract_nxv8f16_v8f16(<8 x half> %data) {
+; CHECK-LABEL: dup_extract_nxv8f16_v8f16:
+; CHECK:       // %bb.0:
+; CHECK-NEXT:    // kill: def $q0 killed $q0 def $z0
+; CHECK-NEXT:    mov z0.h, z0.h[1]
+; CHECK-NEXT:    ret
+  %1 = extractelement <8 x half> %data, i16 1
+  %.splatinsert = insertelement <vscale x 8 x half> poison, half %1, i32 0
+  %.splat = shufflevector <vscale x 8 x half> %.splatinsert, <vscale x 8 x half> poison, <vscale x 8 x i32> zeroinitializer
+  ret <vscale x 8 x half> %.splat
+}
+
+define <vscale x 8 x half> @dup_extract_nxv8f16_v4f16(<4 x half> %data) {
+; CHECK-LABEL: dup_extract_nxv8f16_v4f16:
+; CHECK:       // %bb.0:
+; CHECK-NEXT:    // kill: def $d0 killed $d0 def $z0
+; CHECK-NEXT:    mov z0.h, z0.h[1]
+; CHECK-NEXT:    ret
+  %1 = extractelement <4 x half> %data, i16 1
+  %.splatinsert = insertelement <vscale x 8 x half> poison, half %1, i32 0
+  %.splat = shufflevector <vscale x 8 x half> %.splatinsert, <vscale x 8 x half> poison, <vscale x 8 x i32> zeroinitializer
+  ret <vscale x 8 x half> %.splat
+}
+
+define <vscale x 4 x half> @dup_extract_nxv4f16_nxv8f16(<vscale x 8 x half> %data) {
+; CHECK-LABEL: dup_extract_nxv4f16_nxv8f16:
+; CHECK:       // %bb.0:
+; CHECK-NEXT:    mov z0.h, z0.h[1]
+; CHECK-NEXT:    ret
+  %1 = extractelement <vscale x 8 x half> %data, i16 1
+  %.splatinsert = insertelement <vscale x 4 x half> poison, half %1, i32 0
+  %.splat = shufflevector <vscale x 4 x half> %.splatinsert, <vscale x 4 x half> poison, <vscale x 4 x i32> zeroinitializer
+  ret <vscale x 4 x half> %.splat
+}
+
+define <vscale x 4 x half> @dup_extract_nxv4f16_nxv4f16(<vscale x 4 x half> %data) {
+; CHECK-LABEL: dup_extract_nxv4f16_nxv4f16:
+; CHECK:       // %bb.0:
+; CHECK-NEXT:    mov z0.s, z0.s[1]
+; CHECK-NEXT:    ret
+  %1 = extractelement <vscale x 4 x half> %data, i16 1
+  %.splatinsert = insertelement <vscale x 4 x half> poison, half %1, i32 0
+  %.splat = shufflevector <vscale x 4 x half> %.splatinsert, <vscale x 4 x half> poison, <vscale x 4 x i32> zeroinitializer
+  ret <vscale x 4 x half> %.splat
+}
+
+define <vscale x 4 x half> @dup_extract_nxv4f16_nxv2f16(<vscale x 2 x half> %data) {
+; CHECK-LABEL: dup_extract_nxv4f16_nxv2f16:
+; CHECK:       // %bb.0:
+; CHECK-NEXT:    mov z0.d, z0.d[1]
```

Comment:
This doesn't seem right as we'll only have two copies of the extracted element per 128-bit segment, but we expected four copies, right?

Reply:
I agree, this is definitely broken.

Reply:
Given the original indent was to support extracts from fixed length vectors I've just removed the bogus scalable vector cases.

Reply:
Thanks, it looks good now. Have you considered supporting the remaining unpacked cases (the ones you've just removed) by scaling the index explicitly? For example, for `@dup_extract_nxv4f16_nxv2f16`, I believe we could generate:
```
mov z0.h, z0.h[4]
```
(Instead of the two instructions we currently generate.) I don't think this is too important though.

I am, however, confused about something: the unpacked fixed and scalable extracts seem to index into different underlying elements (see [here](https://godbolt.org/z/r65nj8Mx7) for example). This doesn't directly concern this patch, but I'm struggling to find any references as to why this is the case. Do you happen to have any pointers? Many thanks in advance!

Reply:
Scaling the index requires more tablegen/c++ because I think we'd need a set of ComplexPatterns so I figured I'd do that under a separate PR and keep this one simple by just adding patterns so that I'm not delaying your match patch.

Reply:
Unpacked fixed length vectors don't really exist for AArch64. The fixed length vector types are treated as packed vectors of different length.

Whereas for SVE we treat vectors that are smaller than `vscale x 128` as unpacked vectors whereby there elements are evenly distributed across a `vscale x 128` sized vector.  This is why for scalable vectors the index (or element type) must be scaled but for fixed length they should be used as is.

NOTE: This is for in-reg only.  The memory format of all vector types are considered packed vector of different length.

Reply:
> Scaling the index requires more tablegen/c++ because I think we'd need a set of ComplexPatterns so I figured I'd do that under a separate PR and keep this one simple by just adding patterns so that I'm not delaying your match patch.

That sounds good, thanks very much!

> The reason we do this is to ensure a predicate like `<vscale x N x i1>` works in the same way for all vectors of length `vscale x N` regardless of the element type.

Ah, I see! Cheers, that makes sense!